packages/graphql/src/services/notifications.ts

```diff
@@ -49,3 +52,121 @@ export async function subscribeToPush(
     },
   })
 }
+
+interface UnsubscribeToPushArgs {
+  courseId: string
+  endpoint: string
+}
+
+export async function unsubscribeFromPush( 
+  { courseId, endpoint }: UnsubscribeToPushArgs,
+  ctx: ContextWithUser
+) {
+  try {
+    await ctx.prisma.pushSubscription.delete({
+      where: {
+        participantId_courseId_endpoint: {
+          participantId: ctx.user.sub,
+          courseId,
+          endpoint
+        }
+      }
+    })
+    return true
+  } catch(error) {
+    console.error("An error occured while trying to unsubscribe from push notifications: ", error)
+    return false
+  }
+}
+
+export async function sendPushNotifications( ctx: Context) {
+  webpush.setVapidDetails(
+    'mailto:klicker.support@uzh.ch',
+    process.env.VAPID_PUBLIC_KEY as string,
+    process.env.VAPID_PRIVATE_KEY as string
+  )
+
+  const microSessions = await ctx.prisma.microSession.findMany({
+    where: {
+      status: MicroSessionStatus.PUBLISHED,
+      scheduledStartAt: {
+        lte: new Date(),
+      },
+      scheduledEndAt: {
+        gt: new Date(),
+      },
+      arePushNotificationsSent: false,
+    },
+    include: {
+      course: {
+        include: {
+          subscriptions: true,
+        },
+      },
+      instances: false,
+    },
+  })
+
+  // TODO: improve scalability of push notification dispatch:
+  // 1. Investigate implementing this method as a background process to reduce the load on the main thread.
+  // 2. Investigate implementing this method in Azure
+  await Promise.all(
+    microSessions.map(async (microSession: MicroSession) => {
+      try {
+        await sendPushNotificationsToSubscribers(microSession, ctx);
+        
+        //update microSession to prevent sending push notifications multiple times
+        await ctx.prisma.microSession.update({
+          where: {
+            id: microSession.id,
+          },
+          data: {
+            arePushNotificationsSent: true,
+          },
+        })
+      } catch (error) {
+        console.error("An error occured while trying to send the push notifications: ", error)
+      }
+    })
+ )
+
+  return true
+}
+
+
+async function sendPushNotificationsToSubscribers(microSession: MicroSession, ctx: Context) {
+  for (let sub of microSession.course.subscriptions) {  
+    try {
+      const formattedDate = formatDate(microSession.scheduledEndAt)
+      await webpush.sendNotification(
+        {
+          endpoint: sub.endpoint,
+          keys: {
+            auth: sub.auth,
+            p256dh: sub.p256dh,
+          },
+        },
+        JSON.stringify({
+          message:
+            `Das Microlearning ${microSession.displayName} für ${microSession.course.displayName} ist bis am ${formattedDate.date} um ${formattedDate.time} Uhr verfügbar.`,
```

Comment:
we will need to think about the translation of this when we use multilanguage. do we need to store the language of a course on the course entity and use it here to format the string? what about multilanguage courses? do we store the language on the user subscription (e.g., send it to the backend from the current frontend language when subscribing) and use this language when sending to the specific user?